Userland/Libraries/LibWeb/Layout/BlockFormattingContext.cpp

```diff
@@ -309,18 +306,24 @@ void BlockFormattingContext::compute_width_for_floating_box(Box const& box, Avai
     auto compute_width = [&](auto width) {
         // If 'width' is computed as 'auto', the used value is the "shrink-to-fit" width.
         if (width.is_auto()) {
-
-            // Find the available width: in this case, this is the width of the containing
-            // block minus the used values of 'margin-left', 'border-left-width', 'padding-left',
-            // 'padding-right', 'border-right-width', 'margin-right', and the widths of any relevant scroll bars.
-            auto available_width = width_of_containing_block
-                - margin_left.to_px(box) - computed_values.border_left().width - padding_left
-                - padding_right - computed_values.border_right().width - margin_right.to_px(box);
-
             auto result = calculate_shrink_to_fit_widths(box);
 
-            // Then the shrink-to-fit width is: min(max(preferred minimum width, available width), preferred width).
-            width = CSS::Length::make_px(min(max(result.preferred_minimum_width, available_width), result.preferred_width));
+            if (available_space.width.is_definite()) {
+                // Find the available width: in this case, this is the width of the containing
+                // block minus the used values of 'margin-left', 'border-left-width', 'padding-left',
+                // 'padding-right', 'border-right-width', 'margin-right', and the widths of any relevant scroll bars.
+                auto available_width = available_space.width.to_px_or_zero()
+                    - margin_left.to_px(box) - computed_values.border_left().width - padding_left
+                    - padding_right - computed_values.border_right().width - margin_right.to_px(box);
+                // Then the shrink-to-fit width is: min(max(preferred minimum width, available width), preferred width).
+                width = CSS::Length::make_px(min(max(result.preferred_minimum_width, available_width), result.preferred_width));
+            } else if (available_space.width.is_indefinite() || available_space.width.is_max_content()) {
+                // Fold the formula for shrink-to-fit width for indefinite and max-content available width.
+                width = CSS::Length::make_px(result.preferred_width);
+            } else {
+                // Fold the formula for shrink-to-fit width for min-content available width.
+                width = CSS::Length::make_px(min(result.preferred_minimum_width, result.preferred_width));
+            }
```

Comment:
it looks like the same might be achieved if instead of unwrapping `available_size` to CSSPixels we wrap `preferred_minimum_width` and `preferred_width` to AvailableSize. this should work since we have comparisons operators defined for `AvailableSize` itself.
```suggestion
                // Find the available width: in this case, this is the width of the containing
                // block minus the used values of 'margin-left', 'border-left-width', 'padding-left',
                // 'padding-right', 'border-right-width', 'margin-right', and the widths of any relevant scroll bars.
                auto available_width = available_space.width.to_px_or_zero()
                    - margin_left.to_px(box) - computed_values.border_left().width - padding_left
                    - padding_right - computed_values.border_right().width - margin_right.to_px(box);
                // Then the shrink-to-fit width is: min(max(preferred minimum width, available width), preferred width).
                width = CSS::Length::make_px(min(max(AvailableSize::make_definite(result.preferred_minimum_width), available_width), AvailableSize::make_definite(result.preferred_width)));
```

Reply:
oh, never mind. that's wrong.

Reply:
Thought so too, but `max` wants the same argument type, even if we have comparisons between `AvailableSize` and `CSSPixels`. So we'd have to either wrap `available_width` or roll our own `max` here etc. Unless I misunderstood your suggestion, we'd only be able to sort-of preserve the formula by doing a similar structure of branches.